Keycloak tests failed due to:

```
Caused by: java.lang.ClassNotFoundException: io.quarkus.jackson.ObjectMapperProducer
```

Panache tests failed due to:

```
org.opentest4j.AssertionFailedError: expected: <{"id":null,"name":"max","uniqueName":null,"address":null,"status":null,"dogs":[],"serialisationTrick":1}> but was: <{"id":null,"name":"max","uniqueName":null,"address":null,"status":null,"dogs":[],"serialisationTrick":1,"persistent":false}>
	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
	at org.junit.jupiter.api.AssertionUtils.failNotEqual(AssertionUtils.java:62)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1124)
	at io.quarkus.it.panache.PanacheFunctionalityTest.jacksonDeserilazationIgnoresPersitantAttribute(PanacheFunctionalityTest.java:71)
```

Except the last one is one of those tests that can't possibly fail. It's a test that's reusing a module which doesn't have the fix.

--- a/docs/src/main/asciidoc/hibernate-orm-panache.adoc
+++ b/docs/src/main/asciidoc/hibernate-orm-panache.adoc
@@ -161,6 +161,10 @@ List<Person> allPersons = Person.listAll();
 // finding a specific person by ID
 person = Person.findById(personId);
 
+// finding a specific person by ID via an Optional
+Optional<Person> optional = Person.findByIdOptional(personId);
+person = optional.orElseThrow(() -> NotFoundException());
+
 // finding all living persons
 List<Person> livingPersons = Person.list("status", Status.Alive);
 

--- a/extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/PanacheEntityBase.java
+++ b/extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/PanacheEntityBase.java
@@ -2,6 +2,7 @@ package io.quarkus.hibernate.orm.panache;
 
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.stream.Stream;
 
 import javax.json.bind.annotation.JsonbTransient;
@@ -112,6 +113,29 @@ public abstract class PanacheEntityBase {
      */
     @GenerateBridge(targetReturnTypeErased = true)
     public static <T extends PanacheEntityBase> T findById(Object id, LockModeType lockModeType) {
+        throw JpaOperations.implementationInjectionMissing();
+    }
+
+    /**
+     * Find an entity of this type by ID.
+     *
+     * @param id the ID of the entity to find.
+     * @return if found, an optional containing the entity, else <code>Optional.empty()</code>.
+     */
+    @GenerateBridge
+    public static <T extends PanacheEntityBase> Optional<T> findByIdOptional(Object id) {
+        throw JpaOperations.implementationInjectionMissing();
+    }
+
+    /**
+     * Find an entity of this type by ID.
+     *
+     * @param id the ID of the entity to find.
+     * @param lockModeType the locking strategy to be used when retrieving the entity.
+     * @return if found, an optional containing the entity, else <code>Optional.empty()</code>.
+     */
+    @GenerateBridge
+    public static <T extends PanacheEntityBase> Optional<T> findByIdOptional(Object id, LockModeType lockModeType) {
         throw JpaOperations.implementationInjectionMissing();
     }
 

--- a/extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/PanacheQuery.java
+++ b/extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/PanacheQuery.java
@@ -1,6 +1,7 @@
 package io.quarkus.hibernate.orm.panache;
 
 import java.util.List;
+import java.util.Optional;
 import java.util.stream.Stream;
 
 import javax.persistence.LockModeType;
@@ -164,6 +165,15 @@ public interface PanacheQuery<Entity> {
     public <T extends Entity> T firstResult();
 
     /**
+     * Returns the first result of the current page index. This ignores the current page size to fetch
+     * a single result.
+     *
+     * @return if found, an optional containing the entity, else <code>Optional.empty()</code>.
+     * @see #singleResultOptional()
+     */
+    public <T extends Entity> Optional<T> firstResultOptional();
+
+    /**
      * Executes this query for the current page and return a single result.
      * 
      * @return the single result (throws if there is not exactly one)
@@ -172,4 +182,13 @@ public interface PanacheQuery<Entity> {
      * @see #firstResult()
      */
     public <T extends Entity> T singleResult();
+
+    /**
+     * Executes this query for the current page and return a single result.
+     *
+     * @return if found, an optional containing the entity, else <code>Optional.empty()</code>.
+     * @throws NonUniqueResultException if there are more than one result
+     * @see #firstResultOptional()
+     */
+    public <T extends Entity> Optional<T> singleResultOptional();
 }

--- a/extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/PanacheRepositoryBase.java
+++ b/extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/PanacheRepositoryBase.java
@@ -2,6 +2,7 @@ package io.quarkus.hibernate.orm.panache;
 
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.stream.Stream;
 
 import javax.persistence.LockModeType;
@@ -110,6 +111,28 @@ public interface PanacheRepositoryBase<Entity, Id> {
      */
     @GenerateBridge(targetReturnTypeErased = true)
     public default Entity findById(Id id, LockModeType lockModeType) {
+        throw JpaOperations.implementationInjectionMissing();
+    }
+
+    /**
+     * Find an entity of this type by ID.
+     *
+     * @param id the ID of the entity to find.
+     * @return if found, an optional containing the entity, else <code>Optional.empty()</code>.
+     */
+    @GenerateBridge
+    public default Optional<Entity> findByIdOptional(Id id) {
+        throw JpaOperations.implementationInjectionMissing();
+    }
+
+    /**
+     * Find an entity of this type by ID.
+     *
+     * @param id the ID of the entity to find.
+     * @return if found, an optional containing the entity, else <code>Optional.empty()</code>.
+     */
+    @GenerateBridge
+    public default Optional<Entity> findByIdOptional(Id id, LockModeType lockModeType) {
         throw JpaOperations.implementationInjectionMissing();
     }
 

--- a/extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/JpaOperations.java
+++ b/extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/JpaOperations.java
@@ -3,6 +3,7 @@ package io.quarkus.hibernate.orm.panache.runtime;
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
+import java.util.Optional;
 import java.util.stream.Stream;
 
 import javax.persistence.EntityManager;
@@ -199,6 +200,14 @@ public class JpaOperations {
 
     public static Object findById(Class<?> entityClass, Object id, LockModeType lockModeType) {
         return getEntityManager().find(entityClass, id, lockModeType);
+    }
+
+    public static Optional<?> findByIdOptional(Class<?> entityClass, Object id) {
+        return Optional.ofNullable(findById(entityClass, id));
+    }
+
+    public static Optional<?> findByIdOptional(Class<?> entityClass, Object id, LockModeType lockModeType) {
+        return Optional.ofNullable(findById(entityClass, id, lockModeType));
     }
 
     public static PanacheQuery<?> find(Class<?> entityClass, String query, Object... params) {

--- a/extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java
+++ b/extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java
@@ -2,11 +2,15 @@ package io.quarkus.hibernate.orm.panache.runtime;
 
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.stream.Stream;
 
 import javax.persistence.EntityManager;
 import javax.persistence.LockModeType;
+import javax.persistence.NonUniqueResultException;
 import javax.persistence.Query;
+
+import com.arjuna.ats.internal.jdbc.drivers.modifiers.list;
 
 import io.quarkus.hibernate.orm.panache.PanacheQuery;
 import io.quarkus.panache.common.Page;
@@ -140,9 +144,26 @@ public class PanacheQueryImpl<Entity> implements PanacheQuery<Entity> {
     }
 
     @Override
+    public <T extends Entity> Optional<T> firstResultOptional() {
+        return Optional.ofNullable(firstResult());
+    }
+
+    @Override
     @SuppressWarnings("unchecked")
     public <T extends Entity> T singleResult() {
         jpaQuery.setMaxResults(page.size);
         return (T) jpaQuery.getSingleResult();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T extends Entity> Optional<T> singleResultOptional() {
+        jpaQuery.setMaxResults(2);
+        List<T> list = jpaQuery.getResultList();
+        if (list.size() == 2) {
+            throw new NonUniqueResultException();
+        }
+
+        return list.isEmpty() ? Optional.empty() : Optional.of(list.get(0));
     }
 }

--- a/integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/TestEndpoint.java
+++ b/integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/TestEndpoint.java
@@ -138,7 +138,15 @@ public class TestEndpoint {
         Assertions.assertEquals(person, byId);
         Assertions.assertEquals("Person<" + person.id + ">", byId.toString());
 
+        byId = Person.<Person> findByIdOptional(person.id).get();
+        Assertions.assertEquals(person, byId);
+        Assertions.assertEquals("Person<" + person.id + ">", byId.toString());
+
         byId = Person.findById(person.id, LockModeType.PESSIMISTIC_READ);
+        Assertions.assertEquals(person, byId);
+        Assertions.assertEquals("Person<" + person.id + ">", byId.toString());
+
+        byId = Person.<Person> findByIdOptional(person.id, LockModeType.PESSIMISTIC_READ).get();
         Assertions.assertEquals(person, byId);
         Assertions.assertEquals("Person<" + person.id + ">", byId.toString());
 
@@ -184,6 +192,8 @@ public class TestEndpoint {
         }
 
         Assertions.assertNotNull(Person.findAll().firstResult());
+
+        Assertions.assertNotNull(Person.findAll().firstResultOptional().get());
 
         Assertions.assertEquals(7, Person.deleteAll());
 
@@ -333,7 +343,11 @@ public class TestEndpoint {
         } catch (NoResultException x) {
         }
 
+        Assertions.assertFalse(personDao.findAll().singleResultOptional().isPresent());
+
         Assertions.assertNull(personDao.findAll().firstResult());
+
+        Assertions.assertFalse(personDao.findAll().firstResultOptional().isPresent());
 
         Person person = makeSavedPersonDao();
         Assertions.assertNotNull(person.id);
@@ -363,6 +377,7 @@ public class TestEndpoint {
 
         Assertions.assertEquals(person, personDao.findAll().firstResult());
         Assertions.assertEquals(person, personDao.findAll().singleResult());
+        Assertions.assertEquals(person, personDao.findAll().singleResultOptional().get());
 
         persons = personDao.find("name = ?1", "stef").list();
         Assertions.assertEquals(1, persons.size());
@@ -413,11 +428,18 @@ public class TestEndpoint {
 
         Assertions.assertEquals(person, personDao.find("name", "stef").firstResult());
         Assertions.assertEquals(person, personDao.find("name", "stef").singleResult());
+        Assertions.assertEquals(person, personDao.find("name", "stef").singleResultOptional().get());
 
         Person byId = personDao.findById(person.id);
         Assertions.assertEquals(person, byId);
 
+        byId = personDao.findByIdOptional(person.id).get();
+        Assertions.assertEquals(person, byId);
+
         byId = personDao.findById(person.id, LockModeType.PESSIMISTIC_READ);
+        Assertions.assertEquals(person, byId);
+
+        byId = personDao.findByIdOptional(person.id, LockModeType.PESSIMISTIC_READ).get();
         Assertions.assertEquals(person, byId);
 
         personDao.delete(person);

--- a/integration-tests/mongodb-panache/src/main/java/io/quarkus/it/mongodb/panache/book/BookEntityResource.java
+++ b/integration-tests/mongodb-panache/src/main/java/io/quarkus/it/mongodb/panache/book/BookEntityResource.java
@@ -70,6 +70,12 @@ public class BookEntityResource {
     }
 
     @GET
+    @Path("/optional/{id}")
+    public BookEntity getBookOptional(@PathParam("id") String id) {
+        return BookEntity.<BookEntity> findByIdOptional(new ObjectId(id)).orElseThrow(() -> new NotFoundException());
+    }
+
+    @GET
     @Path("/search/{author}")
     public List<BookEntity> getBooksByAuthor(@PathParam("author") String author) {
         return BookEntity.list("author", author);
@@ -86,7 +92,7 @@ public class BookEntityResource {
         return BookEntity
                 .find("{'creationDate': {$gte: ?1}, 'creationDate': {$lte: ?2}}", LocalDate.parse(dateFrom),
                         LocalDate.parse(dateTo))
-                .firstResult();
+                .<BookEntity> firstResultOptional().orElseThrow(() -> new NotFoundException());
     }
 
     @GET

--- a/integration-tests/mongodb-panache/src/main/java/io/quarkus/it/mongodb/panache/book/BookRepositoryResource.java
+++ b/integration-tests/mongodb-panache/src/main/java/io/quarkus/it/mongodb/panache/book/BookRepositoryResource.java
@@ -73,6 +73,12 @@ public class BookRepositoryResource {
     }
 
     @GET
+    @Path("/optional/{id}")
+    public Book getBookOptional(@PathParam("id") String id) {
+        return bookRepository.findByIdOptional(new ObjectId(id)).orElseThrow(() -> new NotFoundException());
+    }
+
+    @GET
     @Path("/search/{author}")
     public List<Book> getBooksByAuthor(@PathParam("author") String author) {
         return bookRepository.list("author", author);
@@ -89,7 +95,7 @@ public class BookRepositoryResource {
         return bookRepository
                 .find("{'creationDate': {$gte: ?1}, 'creationDate': {$lte: ?2}}", LocalDate.parse(dateFrom),
                         LocalDate.parse(dateTo))
-                .firstResult();
+                .firstResultOptional().orElseThrow(() -> new NotFoundException());
     }
 
     @GET

